File: DSA_Questions/src/com/dsa_questions/Valid_Parenthesis.java
package com.dsa_questions;

import java.util.Stack;

public class Valid_Parenthesis {

public boolean isValid(String s) {
        
        char[] ch = s.toCharArray();
        
        Stack<Character> stack = new Stack<>();
        
        if(ch[0] == ')' || ch[0] == '}' || ch[0] == ']'){
                
                return false;
                
            }
        
        for(int i = 0;i < ch.length;i++){

            
            if(ch[i] == '(' || ch[i] == '{' || ch[i] == '['){
                
                stack.push(ch[i]);
                
            }else{
                 if (stack.isEmpty()) {
                    return false; 
                }
                if(stack.peek() == '(' && ch[i] == ')'){
                    
                    stack.pop();
                    
                    
                }else if(stack.peek() == '{' && ch[i] == '}'){
                    
                    stack.pop();
                    
                }else if(stack.peek() == '[' && ch[i] == ']'){
                    
                    stack.pop();
                    
                }
                else{
                    
                    return false;
                    
                }
                
            }
            
        }
        
        
        if(stack.isEmpty()){
            
            return true;
            
        }else{
            
            return false;
            
        }

    }
	
}

File: DSA_Questions_Python/Valid_Parenthesis.py
class Valid_Parenthesis:
    def isValid(self, s: str) -> bool:
        stack = []

        if not s or s[0] == ')' or s[0] == '}' or s[0] == ']':
            return False

        for ch in s:
            if ch == '(' or ch == '{' or ch == '[':

                stack.append(ch)

            else:

                if not stack:

                    return False
                
                if (stack[-1] == '(' and ch == ')') or \
                   (stack[-1] == '{' and ch == '}') or \
                   (stack[-1] == '[' and ch == ']'):
                    stack.pop()
                else:
                    return False

        return not stack

